dsp/modules/premai.py

```
@@ -0,0 +1,151 @@
+
+        self.project_id = project_id
+        self.session_id = session_id
+
+        if base_url is not None:
+            self.client = premai.Prem(api_key=api_key, base_url=base_url)
+        else:
+            self.client = premai.Prem(api_key=api_key)
+        self.provider = "premai"
+
+        self.kwargs = {
+            "model": model,
+            "temperature": 0.17,
+            "max_tokens": 150,
+            **kwargs,
+        }
+        if session_id is not None:
+            kwargs["sesion_id"] = session_id
+
+    def _get_all_kwargs(self, **kwargs) -> dict:
+        other_kwargs = {
+            "seed": None,
+            "logit_bias": None,
+            "tools": None,
+            "system_prompt": None,
+        }
+        all_kwargs = {
+            **self.kwargs,
+            **other_kwargs,
+            **kwargs,
+        }
+
+        _keys_that_cannot_be_none = [
+            "system_prompt",
+            "frequency_penalty",
+            "presence_penalty",
+            "tools",
+            "model",
+        ]
+
+        for key in _keys_that_cannot_be_none:
+            if all_kwargs.get(key) is None:
+                all_kwargs.pop(key, None)
+        return all_kwargs
+
+    def basic_request(self, prompt, **kwargs) -> str:
+        """Handles retrieval of completions from Prem AI whilst handling API errors."""
+        all_kwargs = self._get_all_kwargs(**kwargs)
+        message = []
+
+        if "system_prompt" in all_kwargs:
+            message.append({"role": "system", "content": all_kwargs["system_prompt"]})
+        message.append({"role": "user", "content": prompt})
+
+        response = self.client.chat.completions.create(
+            project_id=self.project_id,
+            messages=message,
+            **all_kwargs,
+        )
+        if not response.choices:
+            raise ChatPremAPIError("ChatResponse must have at least one candidate")
+
+        return response.choices[0].message.content or ""
```

would it be better to raise this error if the content field is not filled?

---

I mean, yes we can, but we took an assumption that, if `response.choices` exists, then this will also mean that, there should be some response that should exist. If not then there could be any internal error from Prem's side. So, for that we handled it in the above way where if this thing does not exist, then just return ""


what do you think?

---

My comment as more on if `response.choices[0].message.content` is empty, this seems like it would be some incomplete or unexpected response right? and that would be better to raise as a flag over returning `""`.

---

I see, that makes sense. Doing those changes